SeekLowerBound not working correctly

I was trying to replace our custom `Seek` behavior in my fork with `SeekLowerBound` described in here https://github.com/hashicorp/go-immutable-radix/pull/29#issuecomment-627942830 but hit some issues.

First thing I discovered was that an empty key (root) confuses the seeker. Tried to fix this in https://github.com/hashicorp/go-immutable-radix/commit/30c53379998b3a3fc23b26088546e455d5cd25f4

But then I discovered some more cases that behave weirdly. Added tests for them in https://github.com/hashicorp/go-immutable-radix/commit/5bed3aa0fa9f25207cbbe490ebc97ce18aeae6d2

```
    --- FAIL: TestIterateLowerBound/case015 (0.00s)
        iradix_test.go:1712: keys=[ aaa bbb] search=aa want=[aaa bbb]
        iradix_test.go:1730: mis-match: key=aa
              got=[]
              want=[aaa bbb]
    --- FAIL: TestIterateLowerBound/case016 (0.00s)
        iradix_test.go:1712: keys=[ aaa bbb] search=aaab want=[bbb]
        iradix_test.go:1730: mis-match: key=aaab
              got=[]
              want=[bbb]
    --- PASS: TestIterateLowerBound/case017 (0.00s)
        iradix_test.go:1712: keys=[ aaa bbb] search= want=[ aaa bbb]
    --- FAIL: TestIterateLowerBound/case018 (0.00s)
        iradix_test.go:1712: keys=[a aa bbb] search=aa want=[aa bbb]
        iradix_test.go:1730: mis-match: key=aa
              got=[bbb]
              want=[aa bbb]
    --- FAIL: TestIterateLowerBound/case019 (0.00s)
        iradix_test.go:1712: keys=[a aaa bbb] search=aa want=[aaa bbb]
        iradix_test.go:1730: mis-match: key=aa
              got=[bbb]
              want=[aaa bbb]
```

@banks @thaJeztah

--- a/iradix_test.go
+++ b/iradix_test.go
@@ -1735,7 +1735,7 @@ type readableString string
 
 func (s readableString) Generate(rand *rand.Rand, size int) reflect.Value {
 	// Pick a random string from a limited alphabet that makes it easy to read the
-	// failure cases. Also never includes a null byte as we don't support that.
+	// failure cases.
 	const letters = "abcdefghijklmnopqrstuvwxyz/-_0123456789"
 
 	b := make([]byte, size)
@@ -1757,7 +1757,6 @@ func TestIterateLowerBoundFuzz(t *testing.T) {
 	// the same list as filtering all sorted keys that are lower.
 
 	radixAddAndScan := func(newKey, searchKey readableString) []string {
-		// Append a null byte
 		r, _, _ = r.Insert([]byte(newKey), nil)
 
 		// Now iterate the tree from searchKey to the end
@@ -1769,7 +1768,6 @@ func TestIterateLowerBoundFuzz(t *testing.T) {
 			if !ok {
 				break
 			}
-			// Strip the null byte and append to result set
 			result = append(result, string(key))
 		}
 		return result

--- a/iradix_test.go
+++ b/iradix_test.go
@@ -1684,6 +1684,13 @@ func TestIterateLowerBound(t *testing.T) {
 			[]string{"cbbaa", "cbbab", "cbbbc", "cbcbb", "cbcbc", "cbcca", "ccaaa", "ccabc", "ccaca", "ccacc", "ccbac", "cccaa", "cccac", "cccca"},
 		},
 
+		// Panic case found be TestIterateLowerBoundFuzz.
+		{
+			[]string{"gcgc"},
+			"",
+			[]string{"gcgc"},
+		},
+
 		// We SHOULD support keys that are prefixes of each other despite some
 		// confusion in the original implementation.
 		{
@@ -1737,9 +1744,7 @@ func TestIterateLowerBound(t *testing.T) {
 			// Get and seek iterator
 			root := r.Root()
 			iter := root.Iterator()
-			if test.search != "" {
-				iter.SeekLowerBound([]byte(test.search))
-			}
+			iter.SeekLowerBound([]byte(test.search))
 
 			// Consume all the keys
 			out := []string{}
@@ -1763,7 +1768,11 @@ type readableString string
 func (s readableString) Generate(rand *rand.Rand, size int) reflect.Value {
 	// Pick a random string from a limited alphabet that makes it easy to read the
 	// failure cases.
-	const letters = "abcdefghijklmnopqrstuvwxyz/-_0123456789"
+	const letters = "abcdefg"
+
+	// Ignore size and make them all shortish to provoke bigger chance of hitting
+	// prefixes and more intersting tree shapes.
+	size = rand.Intn(8)
 
 	b := make([]byte, size)
 	for i := range b {
@@ -1786,6 +1795,8 @@ func TestIterateLowerBoundFuzz(t *testing.T) {
 	radixAddAndScan := func(newKey, searchKey readableString) []string {
 		r, _, _ = r.Insert([]byte(newKey), nil)
 
+		t.Logf("NewKey: %q, SearchKey: %q", newKey, searchKey)
+
 		// Now iterate the tree from searchKey to the end
 		it := r.Root().Iterator()
 		result := []string{}
@@ -1806,14 +1817,19 @@ func TestIterateLowerBoundFuzz(t *testing.T) {
 		sort.Strings(set)
 
 		t.Logf("Current Set: %#v", set)
+		t.Logf("Search Key: %#v %v", searchKey, "" >= string(searchKey))
 
 		result := []string{}
-		var prev string
-		for _, k := range set {
-			if k >= string(searchKey) && k != prev {
+		for i, k := range set {
+			// Check this is not a duplicate of the previous value. Note we don't just
+			// store the last string to compare because empty string is a valid value
+			// in the set and makes comparing on the first iteration awkward.
+			if i > 0 && set[i-1] == k {
+				continue
+			}
+			if k >= string(searchKey) {
 				result = append(result, k)
 			}
-			prev = k
 		}
 		return result
 	}

--- a/iradix_test.go
+++ b/iradix_test.go
@@ -1532,29 +1532,30 @@ func TestLenTxn(t *testing.T) {
 	}
 }
 
-// these should be defined in order
-var fixedLenKeys = []string{
-	"00000",
-	"00001",
-	"00004",
-	"00010",
-	"00020",
-	"20020",
-}
-
-// these should be defined in order
-var mixedLenKeys = []string{
-	"a1",
-	"abc",
-	"barbazboo",
-	"f",
-	"foo",
-	"found",
-	"zap",
-	"zip",
-}
-
 func TestIterateLowerBound(t *testing.T) {
+
+	// these should be defined in order
+	var fixedLenKeys = []string{
+		"00000",
+		"00001",
+		"00004",
+		"00010",
+		"00020",
+		"20020",
+	}
+
+	// these should be defined in order
+	var mixedLenKeys = []string{
+		"a1",
+		"abc",
+		"barbazboo",
+		"f",
+		"foo",
+		"found",
+		"zap",
+		"zip",
+	}
+
 	type exp struct {
 		keys   []string
 		search string

--- a/iradix_test.go
+++ b/iradix_test.go
@@ -1665,6 +1665,31 @@ func TestIterateLowerBound(t *testing.T) {
 			"cbacb",
 			[]string{"cbbaa", "cbbab", "cbbbc", "cbcbb", "cbcbc", "cbcca", "ccaaa", "ccabc", "ccaca", "ccacc", "ccbac", "cccaa", "cccac", "cccca"},
 		},
+
+		// We SHOULD support keys that are prefixes of each other despite some
+		// confusion in the original implementation.
+		{
+			[]string{"f", "fo", "foo", "food", "bug"},
+			"foo",
+			[]string{"foo", "food"},
+		},
+
+		// We also support the empty key as a valid key to insert and search for.
+		{
+			[]string{"f", "fo", "foo", "food", "bug", ""},
+			"foo",
+			[]string{"foo", "food"},
+		},
+		{
+			[]string{"f", "bug", ""},
+			"",
+			[]string{"", "bug", "f"},
+		},
+		{
+			[]string{"f", "bug", "xylophone"},
+			"",
+			[]string{"bug", "f", "xylophone"},
+		},
 	}
 
 	for idx, test := range cases {
@@ -1725,8 +1750,7 @@ func TestIterateLowerBoundFuzz(t *testing.T) {
 	set := []string{}
 
 	// This specifies a property where each call adds a new random key to the radix
-	// tree (with a null byte appended since our tree doesn't support one key
-	// being a prefix of another and treats null bytes specially).
+	// tree.
 	//
 	// It also maintains a plain sorted list of the same set of keys and asserts
 	// that iterating from some random key to the end using LowerBound produces
@@ -1734,8 +1758,7 @@ func TestIterateLowerBoundFuzz(t *testing.T) {
 
 	radixAddAndScan := func(newKey, searchKey readableString) []string {
 		// Append a null byte
-		key := []byte(newKey + "\x00")
-		r, _, _ = r.Insert(key, nil)
+		r, _, _ = r.Insert([]byte(newKey), nil)
 
 		// Now iterate the tree from searchKey to the end
 		it := r.Root().Iterator()
@@ -1747,7 +1770,7 @@ func TestIterateLowerBoundFuzz(t *testing.T) {
 				break
 			}
 			// Strip the null byte and append to result set
-			result = append(result, string(key[0:len(key)-1]))
+			result = append(result, string(key))
 		}
 		return result
 	}

--- a/iradix_test.go
+++ b/iradix_test.go
@@ -1532,26 +1532,29 @@ func TestLenTxn(t *testing.T) {
 	}
 }
 
+// these should be defined in order
+var fixedLenKeys = []string{
+	"00000",
+	"00001",
+	"00004",
+	"00010",
+	"00020",
+	"20020",
+}
+
+// these should be defined in order
+var mixedLenKeys = []string{
+	"a1",
+	"abc",
+	"barbazboo",
+	"f",
+	"foo",
+	"found",
+	"zap",
+	"zip",
+}
+
 func TestIterateLowerBound(t *testing.T) {
-	fixedLenKeys := []string{
-		"00000",
-		"00001",
-		"00004",
-		"00010",
-		"00020",
-		"20020",
-	}
-
-	mixedLenKeys := []string{
-		"a1",
-		"abc",
-		"barbazboo",
-		"foo",
-		"found",
-		"zap",
-		"zip",
-	}
-
 	type exp struct {
 		keys   []string
 		search string
@@ -1562,7 +1565,8 @@ func TestIterateLowerBound(t *testing.T) {
 			fixedLenKeys,
 			"00000",
 			fixedLenKeys,
-		}, {
+		},
+		{
 			fixedLenKeys,
 			"00003",
 			[]string{
@@ -1571,7 +1575,8 @@ func TestIterateLowerBound(t *testing.T) {
 				"00020",
 				"20020",
 			},
-		}, {
+		},
+		{
 			fixedLenKeys,
 			"00010",
 			[]string{
@@ -1579,64 +1584,77 @@ func TestIterateLowerBound(t *testing.T) {
 				"00020",
 				"20020",
 			},
-		}, {
+		},
+		{
 			fixedLenKeys,
 			"20000",
 			[]string{
 				"20020",
 			},
-		}, {
+		},
+		{
 			fixedLenKeys,
 			"20020",
 			[]string{
 				"20020",
 			},
-		}, {
+		},
+		{
 			fixedLenKeys,
 			"20022",
 			[]string{},
-		}, {
+		},
+		{
 			mixedLenKeys,
 			"A", // before all lower case letters
 			mixedLenKeys,
-		}, {
+		},
+		{
 			mixedLenKeys,
 			"a1",
 			mixedLenKeys,
-		}, {
+		},
+		{
 			mixedLenKeys,
 			"b",
 			[]string{
 				"barbazboo",
+				"f",
 				"foo",
 				"found",
 				"zap",
 				"zip",
 			},
-		}, {
+		},
+		{
 			mixedLenKeys,
 			"bar",
 			[]string{
 				"barbazboo",
+				"f",
 				"foo",
 				"found",
 				"zap",
 				"zip",
 			},
-		}, {
+		},
+		{
 			mixedLenKeys,
 			"barbazboo0",
 			[]string{
+				"f",
 				"foo",
 				"found",
 				"zap",
 				"zip",
 			},
-		}, {
+		},
+		{
 			mixedLenKeys,
 			"zippy",
 			[]string{},
-		}, {
+		},
+		{
 			mixedLenKeys,
 			"zi",
 			[]string{
@@ -1674,7 +1692,8 @@ func TestIterateLowerBound(t *testing.T) {
 			[]string{"foo", "food"},
 		},
 
-		// We also support the empty key as a valid key to insert and search for.
+		// We also support the empty key (which is a prefix of every other key) as a
+		// valid key to insert and search for.
 		{
 			[]string{"f", "fo", "foo", "food", "bug", ""},
 			"foo",
@@ -1689,6 +1708,14 @@ func TestIterateLowerBound(t *testing.T) {
 			[]string{"f", "bug", "xylophone"},
 			"",
 			[]string{"bug", "f", "xylophone"},
+		},
+
+		// This is a case we realized we were not covering while fixing
+		// SeekReverseLowerBound and could panic before.
+		{
+			[]string{"bar", "foo00", "foo11"},
+			"foo",
+			[]string{"foo00", "foo11"},
 		},
 	}
 

--- a/iter.go
+++ b/iter.go
@@ -115,12 +115,8 @@ func (i *Iterator) SeekLowerBound(key []byte) {
 		}
 
 		// Prefix is equal, we are still heading for an exact match. If this is a
-		// leaf we're done.
-		if n.leaf != nil {
-			if bytes.Compare(n.leaf.key, key) < 0 {
-				i.node = nil
-				return
-			}
+		// leaf and an exact match we're done.
+		if n.leaf != nil && bytes.Equal(n.leaf.key, key) {
 			found(n)
 			return
 		}

--- a/iter.go
+++ b/iter.go
@@ -80,13 +80,21 @@ func (i *Iterator) recurseMin(n *Node) *Node {
 func (i *Iterator) SeekLowerBound(key []byte) {
 	// Wipe the stack. Unlike Prefix iteration, we need to build the stack as we
 	// go because we need only a subset of edges of many nodes in the path to the
-	// leaf with the lower bound.
+	// leaf with the lower bound. Note that the iterator will still recurse into
+	// children that we don't traverse on the way to the reverse lower bound as it
+	// walks the stack.
 	i.stack = []edges{}
+	// i.node starts off in the common case as pointing to the root node of the
+	// tree. By the time we return we have either found a lower bound and setup
+	// the stack to traverse all larger keys, or we have not and the stack and
+	// node should both be nil to prevent the iterator from assuming it is just
+	// iterating the whole tree from the root node. Either way this needs to end
+	// up as nil so just set it here.
 	n := i.node
+	i.node = nil
 	search := key
 
 	found := func(n *Node) {
-		i.node = n
 		i.stack = append(i.stack, edges{edge{node: n}})
 	}
 
@@ -96,8 +104,6 @@ func (i *Iterator) SeekLowerBound(key []byte) {
 			found(n)
 			return
 		}
-		i.node = nil
-		return
 	}
 
 	for {
@@ -148,7 +154,6 @@ func (i *Iterator) SeekLowerBound(key []byte) {
 		// Otherwise, take the lower bound next edge.
 		idx, lbNode := n.getLowerBoundEdge(search[0])
 		if lbNode == nil {
-			i.node = nil
 			return
 		}
 
@@ -157,7 +162,6 @@ func (i *Iterator) SeekLowerBound(key []byte) {
 			i.stack = append(i.stack, n.edges[idx+1:])
 		}
 
-		i.node = lbNode
 		// Recurse
 		n = lbNode
 	}

--- a/iter.go
+++ b/iter.go
@@ -60,10 +60,13 @@ func (i *Iterator) recurseMin(n *Node) *Node {
 	if n.leaf != nil {
 		return n
 	}
-	if len(n.edges) > 0 {
+	nEdges := len(n.edges)
+	if nEdges > 1 {
 		// Add all the other edges to the stack (the min node will be added as
 		// we recurse)
 		i.stack = append(i.stack, n.edges[1:])
+	}
+	if nEdges > 0 {
 		return i.recurseMin(n.edges[0].node)
 	}
 	// Shouldn't be possible

--- a/iter.go
+++ b/iter.go
@@ -20,7 +20,7 @@ func (i *Iterator) SeekPrefixWatch(prefix []byte) (watch <-chan struct{}) {
 	watch = n.mutateCh
 	search := prefix
 	for {
-		// Check for key exhaution
+		// Check for key exhaustion
 		if len(search) == 0 {
 			i.node = n
 			return
@@ -87,6 +87,16 @@ func (i *Iterator) SeekLowerBound(key []byte) {
 		i.stack = append(i.stack, edges{edge{node: n}})
 	}
 
+	findMin := func(n *Node) {
+		n = i.recurseMin(n)
+		if n != nil {
+			found(n)
+			return
+		}
+		i.node = nil
+		return
+	}
+
 	for {
 		// Compare current prefix with the search key's same-length prefix.
 		var prefixCmp int
@@ -100,10 +110,7 @@ func (i *Iterator) SeekLowerBound(key []byte) {
 			// Prefix is larger, that means the lower bound is greater than the search
 			// and from now on we need to follow the minimum path to the smallest
 			// leaf under this subtree.
-			n = i.recurseMin(n)
-			if n != nil {
-				found(n)
-			}
+			findMin(n)
 			return
 		}
 
@@ -121,11 +128,18 @@ func (i *Iterator) SeekLowerBound(key []byte) {
 			return
 		}
 
-		// Consume the search prefix
-		if len(n.prefix) > len(search) {
-			search = []byte{}
-		} else {
-			search = search[len(n.prefix):]
+		// Consume the search prefix if the current node has one. Note that this is
+		// safe because if n.prefix is longer than the search slice prefixCmp would
+		// have been > 0 above and the method would have already returned.
+		search = search[len(n.prefix):]
+
+		if len(search) == 0 {
+			// We've exhausted the search key, but the current node is not an exact
+			// match or not a leaf. That means that the leaf value if it exists, and
+			// all child nodes must be strictly greater, the smallest key in this
+			// subtree must be the lower bound.
+			findMin(n)
+			return
 		}
 
 		// Otherwise, take the lower bound next edge.

--- a/reverse_iter.go
+++ b/reverse_iter.go
@@ -9,20 +9,14 @@ import (
 type ReverseIterator struct {
 	i *Iterator
 
-	// Unlike forward iteration we need to track parent nodes that have already
-	// had their children pushed to the stack because internal nodes may also be
-	// leaves but we need to return all of their children (who must be greater
-	// than them) before we can return them. If we just pushed them back onto the
-	// stack without marking them in some other way we'd end up in an infinite
-	// loop, if we don't push them back we'd end up skipping them, or skipping all
-	// their children. This map stores the set of parent nodes that we have
-	// descended through while walking the stack but that still need to have their
-	// leaf values returned when we are done with all their children.
+	// expandedParents stores the set of parent nodes whose relevant children have
+	// already been pushed into the stack. This can happen during seek or during
+	// iteration.
 	//
-	// SeekReverseLowerBound also has to set this on internal leaves it traverses
-	// through since it will already be adding any relevant children it finds to
-	// the stack so the iterator shouldn't walk into it's children again when it
-	// gets to it in the stack.
+	// Unlike forward iteration we need to recurse into children before we can
+	// output the value stored in an internal leaf since all children are greater.
+	// We use this to track whether we have already ensured all the children are
+	// in the stack.
 	expandedParents map[*Node]struct{}
 }
 
@@ -118,7 +112,10 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 			return
 		}
 
-		// If this is a leaf, something needs to happen!
+		// If this is a leaf, something needs to happen! Note that if it's a leaf
+		// and prefixCmp was zero (which it must be to get here) then the leaf value
+		// is either an exact match for the search, or it's lower. It can't be
+		// greater.
 		if n.isLeaf() {
 
 			// Firstly, if it's an exact match, we're done!
@@ -127,26 +124,20 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 				return
 			}
 
-			// If not exact, and it has no children then we are also done, decide if
-			// it's the lowerBound of if none exists.
+			// It's not so this node's leaf value must be lower and could still be a
+			// valide contender for reverse lower bound.
+
+			// If it has no children then we are also done.
 			if len(n.edges) == 0 {
-				if bytes.Compare(n.leaf.key, key) > 0 {
-					// This leaf is greater than the search key so no reverse lower bound
-					// exists in this subtree. There may be one in another leaf to the
-					// left of the path we took which will already be in the iterator's
-					// stack.
-					ri.i.node = nil
-					return
-				}
-				// this leaf is the lower bound.
+				// This leaf is the lower bound.
 				found(n)
 				return
 			}
 
-			// Finally, this leaf is internal and has children so we'll keep
-			// searching, but we need to add it to the iterator's stack since it has a
-			// leaf value that needs to be iterated over. It needs to be added to the
-			// stack before it's children below as it comes first.
+			// Finally, this leaf is internal (has children) so we'll keep searching,
+			// but we need to add it to the iterator's stack since it has a leaf value
+			// that needs to be iterated over. It needs to be added to the stack
+			// before it's children below as it comes first.
 			ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
 			// We also need to mark it as expanded since we'll be adding any of it's
 			// relevant children below and so don't want the iterator to re-add them

--- a/reverse_iter.go
+++ b/reverse_iter.go
@@ -66,9 +66,9 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 	found := func(n *Node) {
 		ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
 		// We need to mark this node as expanded in advance too otherwise the
-		// iterator will attempt to walk all of it's children even though they are
+		// iterator will attempt to walk all of its children even though they are
 		// greater than the lower bound we have found. We've expanded it in the
-		// sense that all of it's children that we want to walk are already in the
+		// sense that all of its children that we want to walk are already in the
 		// stack (i.e. none of them).
 		ri.expandedParents[n] = struct{}{}
 	}
@@ -90,7 +90,7 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 			// tree. So we need to follow the maximum path in this subtree to find it.
 			// Note that this is exactly what the iterator will already do if it find
 			// a node in the stack that has _not_ been marked as expanded so in this
-			// one case we don't call `found` and just let the iterator do it's
+			// one case we don't call `found` and just let the iterator do its
 			// expansion and recursion through all the children.
 			ri.i.node = n
 			ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
@@ -118,7 +118,7 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 			}
 
 			// It's not so this node's leaf value must be lower and could still be a
-			// valide contender for reverse lower bound.
+			// valid contender for reverse lower bound.
 
 			// If it has no children then we are also done.
 			if len(n.edges) == 0 {
@@ -130,11 +130,11 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 			// Finally, this leaf is internal (has children) so we'll keep searching,
 			// but we need to add it to the iterator's stack since it has a leaf value
 			// that needs to be iterated over. It needs to be added to the stack
-			// before it's children below as it comes first.
+			// before its children below as it comes first.
 			ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
-			// We also need to mark it as expanded since we'll be adding any of it's
+			// We also need to mark it as expanded since we'll be adding any of its
 			// relevant children below and so don't want the iterator to re-add them
-			// on it's way back up the stack.
+			// on its way back up the stack.
 			ri.expandedParents[n] = struct{}{}
 		}
 
@@ -234,7 +234,7 @@ func (ri *ReverseIterator) Previous() ([]byte, interface{}, bool) {
 			return elem.leaf.key, elem.leaf.val, true
 		}
 
-		// it's not a leaf so keep walking the stack to find the previous leafq
+		// it's not a leaf so keep walking the stack to find the previous leaf
 	}
 	return nil, nil, false
 }

--- a/reverse_iter.go
+++ b/reverse_iter.go
@@ -45,9 +45,18 @@ func (ri *ReverseIterator) SeekPrefix(prefix []byte) {
 func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 	// Wipe the stack. Unlike Prefix iteration, we need to build the stack as we
 	// go because we need only a subset of edges of many nodes in the path to the
-	// leaf with the lower bound.
+	// leaf with the lower bound. Note that the iterator will still recurse into
+	// children that we don't traverse on the way to the reverse lower bound as it
+	// walks the stack.
 	ri.i.stack = []edges{}
+	// ri.i.node starts off in the common case as pointing to the root node of the
+	// tree. By the time we return we have either found a lower bound and setup
+	// the stack to traverse all larger keys, or we have not and the stack and
+	// node should both be nil to prevent the iterator from assuming it is just
+	// iterating the whole tree from the root node. Either way this needs to end
+	// up as nil so just set it here.
 	n := ri.i.node
+	ri.i.node = nil
 	search := key
 
 	if ri.expandedParents == nil {
@@ -55,7 +64,6 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 	}
 
 	found := func(n *Node) {
-		ri.i.node = n
 		ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
 		// We need to mark this node as expanded in advance too otherwise the
 		// iterator will attempt to walk all of it's children even though they are
@@ -94,7 +102,6 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 			// also exhausted the search key. Either way, that means there is no
 			// reverse lower bound since nothing comes before our current search
 			// prefix.
-			ri.i.node = nil
 			return
 		}
 
@@ -143,7 +150,6 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 			// the whole radix tree by following a different path somewhere further
 			// up. If that's the case then the iterator's stack will contain all the
 			// smaller nodes already and Previous will walk through them correctly.
-			ri.i.node = nil
 			return
 		}
 
@@ -167,11 +173,9 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 		// Exit if there's no lower bound edge. The stack will have the previous
 		// nodes already.
 		if lbNode == nil {
-			ri.i.node = nil
 			return
 		}
 
-		ri.i.node = lbNode
 		// Recurse
 		n = lbNode
 	}

--- a/reverse_iter.go
+++ b/reverse_iter.go
@@ -83,16 +83,15 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 		}
 
 		if prefixCmp < 0 {
-			// Prefix is smaller than search prefix, that means there is no lower
-			// bound. But we are looking in reverse, so the reverse lower bound will
-			// be the largest leaf under this subtree, since it is the value that
-			// would come right before the current search prefix if it were in the
-			// tree. So we need to follow the maximum path in this subtree to find it.
-			// Note that this is exactly what the iterator will already do if it find
-			// a node in the stack that has _not_ been marked as expanded so in this
-			// one case we don't call `found` and just let the iterator do its
-			// expansion and recursion through all the children.
-			ri.i.node = n
+			// Prefix is smaller than search prefix, that means there is no exact
+			// match for the search key. But we are looking in reverse, so the reverse
+			// lower bound will be the largest leaf under this subtree, since it is
+			// the value that would come right before the current search key if it
+			// were in the tree. So we need to follow the maximum path in this subtree
+			// to find it. Note that this is exactly what the iterator will already do
+			// if it finds a node in the stack that has _not_ been marked as expanded
+			// so in this one case we don't call `found` and instead let the iterator
+			// do the expansion and recursion through all the children.
 			ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
 			return
 		}

--- a/reverse_iter.go
+++ b/reverse_iter.go
@@ -244,7 +244,7 @@ func (ri *ReverseIterator) Previous() ([]byte, interface{}, bool) {
 			return elem.leaf.key, elem.leaf.val, true
 		}
 
-		// it's not a leaf so keep walking the stack to find the previous leaf
+		// it's not a leaf so keep walking the stack to find the previous leafq
 	}
 	return nil, nil, false
 }

--- a/reverse_iter.go
+++ b/reverse_iter.go
@@ -8,6 +8,22 @@ import (
 // in reverse in-order
 type ReverseIterator struct {
 	i *Iterator
+
+	// Unlike forward iteration we need to track parent nodes that have already
+	// had their children pushed to the stack because internal nodes may also be
+	// leaves but we need to return all of their children (who must be greater
+	// than them) before we can return them. If we just pushed them back onto the
+	// stack without marking them in some other way we'd end up in an infinite
+	// loop, if we don't push them back we'd end up skipping them, or skipping all
+	// their children. This map stores the set of parent nodes that we have
+	// descended through while walking the stack but that still need to have their
+	// leaf values returned when we are done with all their children.
+	//
+	// SeekReverseLowerBound also has to set this on internal leaves it traverses
+	// through since it will already be adding any relevant children it finds to
+	// the stack so the iterator shouldn't walk into it's children again when it
+	// gets to it in the stack.
+	expandedParents map[*Node]struct{}
 }
 
 // NewReverseIterator returns a new ReverseIterator at a node
@@ -56,9 +72,19 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 	n := ri.i.node
 	search := key
 
+	if ri.expandedParents == nil {
+		ri.expandedParents = make(map[*Node]struct{})
+	}
+
 	found := func(n *Node) {
 		ri.i.node = n
 		ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
+		// We need to mark this node as expanded in advance too otherwise the
+		// iterator will attempt to walk all of it's children even though they are
+		// greater than the lower bound we have found. We've expanded it in the
+		// sense that all of it's children that we want to walk are already in the
+		// stack (i.e. none of them).
+		ri.expandedParents[n] = struct{}{}
 	}
 
 	for {
@@ -84,28 +110,64 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 		}
 
 		if prefixCmp > 0 {
-			// Prefix is larger than search prefix, that means there is no reverse lower
-			// bound since nothing comes before our current search prefix.
+			// Prefix is larger than search prefix, or there is no prefix but we've
+			// also exhausted the search key. Either way, that means there is no
+			// reverse lower bound since nothing comes before our current search
+			// prefix.
 			ri.i.node = nil
 			return
 		}
 
-		// Prefix is equal, we are still heading for an exact match. If this is a
-		// leaf we're done.
-		if n.leaf != nil {
-			if bytes.Compare(n.leaf.key, key) < 0 {
-				ri.i.node = nil
+		// If this is a leaf, something needs to happen!
+		if n.isLeaf() {
+
+			// Firstly, if it's an exact match, we're done!
+			if bytes.Equal(n.leaf.key, key) {
+				found(n)
 				return
 			}
-			found(n)
-			return
+
+			// If not exact, and it has no children then we are also done, decide if
+			// it's the lowerBound of if none exists.
+			if len(n.edges) == 0 {
+				if bytes.Compare(n.leaf.key, key) > 0 {
+					// This leaf is greater than the search key so no reverse lower bound
+					// exists in this subtree. There may be one in another leaf to the
+					// left of the path we took which will already be in the iterator's
+					// stack.
+					ri.i.node = nil
+					return
+				}
+				// this leaf is the lower bound.
+				found(n)
+				return
+			}
+
+			// Finally, this leaf is internal and has children so we'll keep
+			// searching, but we need to add it to the iterator's stack since it has a
+			// leaf value that needs to be iterated over. It needs to be added to the
+			// stack before it's children below as it comes first.
+			ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
+			// We also need to mark it as expanded since we'll be adding any of it's
+			// relevant children below and so don't want the iterator to re-add them
+			// on it's way back up the stack.
+			ri.expandedParents[n] = struct{}{}
 		}
 
-		// Consume the search prefix
-		if len(n.prefix) > len(search) {
-			search = []byte{}
-		} else {
-			search = search[len(n.prefix):]
+		// Consume the search prefix. Note that this is safe because if n.prefix is
+		// longer than the search slice prefixCmp would have been > 0 above and the
+		// method would have already returned.
+		search = search[len(n.prefix):]
+
+		if len(search) == 0 {
+			// We've exhausted the search key but we are not at a leaf. That means all
+			// children are greater than the search key so a reverse lower bound
+			// doesn't exist in this subtree. Note that there might still be one in
+			// the whole radix tree by following a different path somewhere further
+			// up. If that's the case then the iterator's stack will contain all the
+			// smaller nodes already and Previous will walk through them correctly.
+			ri.i.node = nil
+			return
 		}
 
 		// Otherwise, take the lower bound next edge.
@@ -125,8 +187,8 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 			ri.i.stack = append(ri.i.stack, n.edges[:idx])
 		}
 
-		// Exit if there's not lower bound edge. The stack will have the
-		// previous nodes already.
+		// Exit if there's no lower bound edge. The stack will have the previous
+		// nodes already.
 		if lbNode == nil {
 			ri.i.node = nil
 			return
@@ -149,6 +211,10 @@ func (ri *ReverseIterator) Previous() ([]byte, interface{}, bool) {
 		}
 	}
 
+	if ri.expandedParents == nil {
+		ri.expandedParents = make(map[*Node]struct{})
+	}
+
 	for len(ri.i.stack) > 0 {
 		// Inspect the last element of the stack
 		n := len(ri.i.stack)
@@ -156,22 +222,38 @@ func (ri *ReverseIterator) Previous() ([]byte, interface{}, bool) {
 		m := len(last)
 		elem := last[m-1].node
 
-		// Update the stack
+		_, alreadyExpanded := ri.expandedParents[elem]
+
+		// If this is an internal node and we've not seen it already, we need to
+		// leave it in the stack so we can return its posssible leaf value _after_
+		// we've recursed through all it's children.
+		if len(elem.edges) > 0 && !alreadyExpanded {
+			// record that we've seen this node!
+			ri.expandedParents[elem] = struct{}{}
+			// push child edges onto stack and skip the rest of the loop to recurse
+			// into the largest one.
+			ri.i.stack = append(ri.i.stack, elem.edges)
+			continue
+		}
+
+		// Remove the node from the stack
 		if m > 1 {
 			ri.i.stack[n-1] = last[:m-1]
 		} else {
 			ri.i.stack = ri.i.stack[:n-1]
 		}
-
-		// Push the edges onto the frontier
-		if len(elem.edges) > 0 {
-			ri.i.stack = append(ri.i.stack, elem.edges)
+		// We don't need this state any more as it's no longer in the stack so we
+		// won't visit it again
+		if alreadyExpanded {
+			delete(ri.expandedParents, elem)
 		}
 
-		// Return the leaf values if any
+		// If this is a leaf, return it
 		if elem.leaf != nil {
 			return elem.leaf.key, elem.leaf.val, true
 		}
+
+		// it's not a leaf so keep walking the stack to find the previous leaf
 	}
 	return nil, nil, false
 }

--- a/reverse_iter.go
+++ b/reverse_iter.go
@@ -38,22 +38,6 @@ func (ri *ReverseIterator) SeekPrefix(prefix []byte) {
 	ri.i.SeekPrefixWatch(prefix)
 }
 
-func (ri *ReverseIterator) recurseMax(n *Node) *Node {
-	// Traverse to the maximum child
-	if n.leaf != nil {
-		return n
-	}
-	if len(n.edges) > 0 {
-		// Add all the other edges to the stack (the max node will be added as
-		// we recurse)
-		m := len(n.edges)
-		ri.i.stack = append(ri.i.stack, n.edges[:m-1])
-		return ri.recurseMax(n.edges[m-1].node)
-	}
-	// Shouldn't be possible
-	return nil
-}
-
 // SeekReverseLowerBound is used to seek the iterator to the largest key that is
 // lower or equal to the given key. There is no watch variant as it's hard to
 // predict based on the radix structure which node(s) changes might affect the
@@ -91,15 +75,17 @@ func (ri *ReverseIterator) SeekReverseLowerBound(key []byte) {
 		}
 
 		if prefixCmp < 0 {
-			// Prefix is smaller than search prefix, that means there is no lower bound.
-			// But we are looking in reverse, so the reverse lower bound will be the
-			// largest leaf under this subtree, since it is the value that would come
-			// right before the current search prefix if it were in the tree. So we need
-			// to follow the maximum path in this subtree to find it.
-			n = ri.recurseMax(n)
-			if n != nil {
-				found(n)
-			}
+			// Prefix is smaller than search prefix, that means there is no lower
+			// bound. But we are looking in reverse, so the reverse lower bound will
+			// be the largest leaf under this subtree, since it is the value that
+			// would come right before the current search prefix if it were in the
+			// tree. So we need to follow the maximum path in this subtree to find it.
+			// Note that this is exactly what the iterator will already do if it find
+			// a node in the stack that has _not_ been marked as expanded so in this
+			// one case we don't call `found` and just let the iterator do it's
+			// expansion and recursion through all the children.
+			ri.i.node = n
+			ri.i.stack = append(ri.i.stack, edges{edge{node: n}})
 			return
 		}
 
@@ -216,8 +202,8 @@ func (ri *ReverseIterator) Previous() ([]byte, interface{}, bool) {
 		_, alreadyExpanded := ri.expandedParents[elem]
 
 		// If this is an internal node and we've not seen it already, we need to
-		// leave it in the stack so we can return its posssible leaf value _after_
-		// we've recursed through all it's children.
+		// leave it in the stack so we can return its possible leaf value _after_
+		// we've recursed through all its children.
 		if len(elem.edges) > 0 && !alreadyExpanded {
 			// record that we've seen this node!
 			ri.expandedParents[elem] = struct{}{}

--- a/reverse_iter_test.go
+++ b/reverse_iter_test.go
@@ -1,6 +1,8 @@
 package iradix
 
 import (
+	"fmt"
+	"reflect"
 	"sort"
 	"testing"
 	"testing/quick"
@@ -67,7 +69,7 @@ func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
 	// tree are lower than the search key) and the seek process is cut short.
 	//
 	// But when starting from a non-root node, the prefix is not empty and so
-	// it will require a recursive search for the glabal maximum in the
+	// it will require a recursive search for the global maximum in the
 	// sub-tree, which is not needed when starting from the root.
 
 	r := New()
@@ -121,6 +123,187 @@ func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
 
 	if err := quick.CheckEqual(radixAddAndScan, sliceAddSortAndFilter, nil); err != nil {
 		t.Error(err)
+	}
+}
+
+func TestReverseIterator_SeekLowerBound(t *testing.T) {
+
+	var revFixedLenKeys, revMixedLenKeys []string
+
+	for i := len(fixedLenKeys) - 1; i >= 0; i-- {
+		revFixedLenKeys = append(revFixedLenKeys, fixedLenKeys[i])
+	}
+	for i := len(mixedLenKeys) - 1; i >= 0; i-- {
+		revMixedLenKeys = append(revMixedLenKeys, mixedLenKeys[i])
+	}
+
+	type exp struct {
+		keys   []string
+		search string
+		want   []string
+	}
+	cases := []exp{
+		{
+			fixedLenKeys,
+			"20020",
+			revFixedLenKeys,
+		},
+		{
+			fixedLenKeys,
+			"20000",
+			[]string{
+				"00020",
+				"00010",
+				"00004",
+				"00001",
+				"00000",
+			},
+		},
+		{
+			fixedLenKeys,
+			"00010",
+			[]string{
+				"00010",
+				"00004",
+				"00001",
+				"00000",
+			},
+		},
+		{
+			fixedLenKeys,
+			"00000",
+			[]string{
+				"00000",
+			},
+		},
+		{
+			fixedLenKeys,
+			"0",
+			[]string{},
+		},
+		{
+			mixedLenKeys,
+			"{", // after all lower case letters
+			revMixedLenKeys,
+		},
+		{
+			mixedLenKeys,
+			"zip",
+			revMixedLenKeys,
+		},
+		{
+			mixedLenKeys,
+			"b",
+			[]string{
+				"abc",
+				"a1",
+			},
+		},
+		{
+			mixedLenKeys,
+			"barbazboo0",
+			[]string{
+				"barbazboo",
+				"abc",
+				"a1",
+			},
+		},
+		{
+			mixedLenKeys,
+			"a",
+			[]string{},
+		},
+		{
+			mixedLenKeys,
+			"a1",
+			[]string{
+				"a1",
+			},
+		},
+
+		// We SHOULD support keys that are prefixes of each other despite some
+		// confusion in the original implementation.
+		{
+			[]string{"f", "fo", "foo", "food", "bug"},
+			"foo",
+			[]string{"foo", "fo", "f", "bug"},
+		},
+		{
+			[]string{"f", "fo", "foo", "food", "bug"},
+			"foozzzzzzzzzz", // larger than any key but with shared prefix
+			[]string{"food", "foo", "fo", "f", "bug"},
+		},
+
+		// We also support the empty key (which is a prefix of every other key) as a
+		// valid key to insert and search for.
+		{
+			[]string{"f", "fo", "foo", "food", "bug", ""},
+			"foo",
+			[]string{"foo", "fo", "f", "bug", ""},
+		},
+		{
+			[]string{"f", "bug", ""},
+			"",
+			[]string{""},
+		},
+		{
+			[]string{"f", "bug", "xylophone"},
+			"",
+			[]string{},
+		},
+
+		// This case could panic before. it involves a node with a shared prefix and
+		// children where the reverse lower bound is greater than all the children
+		{
+			[]string{"foo00", "foo11"},
+			"foo",
+			[]string{},
+		},
+
+		// When fixing the panic above the above test could pass but we need to
+		// verify the logic is still correct in the case there was a lower bound in
+		// another node.
+		{
+			[]string{"bar", "foo00", "foo11"},
+			"foo",
+			[]string{"bar"},
+		},
+	}
+
+	for idx, test := range cases {
+		t.Run(fmt.Sprintf("case%03d", idx), func(t *testing.T) {
+			r := New()
+
+			// Insert keys
+			for _, k := range test.keys {
+				var ok bool
+				r, _, ok = r.Insert([]byte(k), nil)
+				if ok {
+					t.Fatalf("duplicate key %s in keys", k)
+				}
+			}
+			if r.Len() != len(test.keys) {
+				t.Fatal("failed adding keys")
+			}
+			// Get and seek iterator
+			root := r.Root()
+			iter := root.ReverseIterator()
+			iter.SeekReverseLowerBound([]byte(test.search))
+
+			// Consume all the keys
+			out := []string{}
+			for {
+				key, _, ok := iter.Previous()
+				if !ok {
+					break
+				}
+				out = append(out, string(key))
+			}
+			if !reflect.DeepEqual(out, test.want) {
+				t.Fatalf("mis-match: key=%s\n  got=%v\n  want=%v", test.search,
+					out, test.want)
+			}
+		})
 	}
 }
 

--- a/reverse_iter_test.go
+++ b/reverse_iter_test.go
@@ -41,14 +41,22 @@ func TestReverseIterator_SeekReverseLowerBoundFuzz(t *testing.T) {
 		set = append(set, string(newKey))
 		sort.Strings(set)
 
+		t.Logf("Current Set: %#v", set)
+		t.Logf("Search Key: %#v %v", searchKey, "" >= string(searchKey))
+
 		result := []string{}
-		var prev string
 		for i := len(set) - 1; i >= 0; i-- {
 			k := set[i]
-			if k <= string(searchKey) && k != prev {
+			// Check this is not a duplicate of the previous value we just included.
+			// Note we don't just store the last string to compare because empty
+			// string is a valid value in the set and makes comparing on the first
+			// iteration awkward.
+			if i < len(set)-2 && set[i+1] == k {
+				continue
+			}
+			if k <= string(searchKey) {
 				result = append(result, k)
 			}
-			prev = k
 		}
 		return result
 	}
@@ -110,13 +118,18 @@ func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
 		sort.Strings(set)
 
 		result := []string{}
-		var prev string
 		for i := len(set) - 1; i >= 0; i-- {
 			k := set[i]
-			if k <= string(searchKey) && k[:len(n.prefix)] <= string(n.prefix) && k != prev {
+			// Check this is not a duplicate of the previous value we just included.
+			// Note we don't just store the last string to compare because empty
+			// string is a valid value in the set and makes comparing on the first
+			// iteration awkward.
+			if i < len(set)-2 && set[i+1] == k {
+				continue
+			}
+			if k <= string(searchKey) && k[:len(n.prefix)] <= string(n.prefix) {
 				result = append(result, k)
 			}
-			prev = k
 		}
 		return result
 	}
@@ -267,6 +280,14 @@ func TestReverseIterator_SeekLowerBound(t *testing.T) {
 			[]string{"bar", "foo00", "foo11"},
 			"foo",
 			[]string{"bar"},
+		},
+
+		// Found by fuzz test to hit code that wasn't covered by any other example
+		// here.
+		{
+			[]string{"bdgedcdc", "agcbcaba"},
+			"beefdafg",
+			[]string{"bdgedcdc", "agcbcaba"},
 		},
 	}
 

--- a/reverse_iter_test.go
+++ b/reverse_iter_test.go
@@ -68,13 +68,26 @@ func TestReverseIterator_SeekReverseLowerBoundFuzz(t *testing.T) {
 
 func TestReverseIterator_SeekLowerBound(t *testing.T) {
 
-	var revFixedLenKeys, revMixedLenKeys []string
-
-	for i := len(fixedLenKeys) - 1; i >= 0; i-- {
-		revFixedLenKeys = append(revFixedLenKeys, fixedLenKeys[i])
+	// these should be defined in order
+	var fixedLenKeys = []string{
+		"20020",
+		"00020",
+		"00010",
+		"00004",
+		"00001",
+		"00000",
 	}
-	for i := len(mixedLenKeys) - 1; i >= 0; i-- {
-		revMixedLenKeys = append(revMixedLenKeys, mixedLenKeys[i])
+
+	// these should be defined in order
+	var mixedLenKeys = []string{
+		"zip",
+		"zap",
+		"found",
+		"foo",
+		"f",
+		"barbazboo",
+		"abc",
+		"a1",
 	}
 
 	type exp struct {
@@ -86,7 +99,7 @@ func TestReverseIterator_SeekLowerBound(t *testing.T) {
 		{
 			fixedLenKeys,
 			"20020",
-			revFixedLenKeys,
+			fixedLenKeys,
 		},
 		{
 			fixedLenKeys,
@@ -124,12 +137,12 @@ func TestReverseIterator_SeekLowerBound(t *testing.T) {
 		{
 			mixedLenKeys,
 			"{", // after all lower case letters
-			revMixedLenKeys,
+			mixedLenKeys,
 		},
 		{
 			mixedLenKeys,
 			"zip",
-			revMixedLenKeys,
+			mixedLenKeys,
 		},
 		{
 			mixedLenKeys,

--- a/reverse_iter_test.go
+++ b/reverse_iter_test.go
@@ -11,17 +11,14 @@ func TestReverseIterator_SeekReverseLowerBoundFuzz(t *testing.T) {
 	set := []string{}
 
 	// This specifies a property where each call adds a new random key to the radix
-	// tree (with a null byte appended since our tree doesn't support one key
-	// being a prefix of another and treats null bytes specially).
+	// tree.
 	//
 	// It also maintains a plain sorted list of the same set of keys and asserts
 	// that iterating from some random key to the beginning using ReverseLowerBound
 	// produces the same list as filtering all sorted keys that are bigger.
 
 	radixAddAndScan := func(newKey, searchKey readableString) []string {
-		// Append a null byte
-		key := []byte(newKey + "\x00")
-		r, _, _ = r.Insert(key, nil)
+		r, _, _ = r.Insert([]byte(newKey), nil)
 
 		// Now iterate the tree from searchKey to the beginning
 		it := r.Root().ReverseIterator()
@@ -32,8 +29,7 @@ func TestReverseIterator_SeekReverseLowerBoundFuzz(t *testing.T) {
 			if !ok {
 				break
 			}
-			// Strip the null byte and append to result set
-			result = append(result, string(key[0:len(key)-1]))
+			result = append(result, string(key))
 		}
 		return result
 	}
@@ -79,9 +75,7 @@ func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
 	var n *Node
 
 	radixAddAndScan := func(newKey, searchKey readableString) []string {
-		// Append a null byte
-		key := []byte(newKey + "\x00")
-		r, _, _ = r.Insert(key, nil)
+		r, _, _ = r.Insert([]byte(newKey), nil)
 
 		// Start seeking from the first root child or don't seek yet
 		if len(r.Root().edges) == 0 {
@@ -98,8 +92,7 @@ func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
 			if !ok {
 				break
 			}
-			// Strip the null byte and append to result set
-			result = append(result, string(key[0:len(key)-1]))
+			result = append(result, string(key))
 		}
 		return result
 	}
@@ -110,12 +103,6 @@ func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
 			return []string{}
 		}
 
-		// Remove the null byte from the prefix if present
-		prefix := n.prefix
-		if prefix[len(prefix)-1] == byte('\x00') {
-			prefix = prefix[:len(prefix)-1]
-		}
-
 		// Append the key to the set and re-sort
 		set = append(set, string(newKey))
 		sort.Strings(set)
@@ -124,7 +111,7 @@ func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
 		var prev string
 		for i := len(set) - 1; i >= 0; i-- {
 			k := set[i]
-			if k <= string(searchKey) && k[:len(prefix)] <= string(prefix) && k != prev {
+			if k <= string(searchKey) && k[:len(n.prefix)] <= string(n.prefix) && k != prev {
 				result = append(result, k)
 			}
 			prev = k

--- a/reverse_iter_test.go
+++ b/reverse_iter_test.go
@@ -51,83 +51,10 @@ func TestReverseIterator_SeekReverseLowerBoundFuzz(t *testing.T) {
 			// Note we don't just store the last string to compare because empty
 			// string is a valid value in the set and makes comparing on the first
 			// iteration awkward.
-			if i < len(set)-2 && set[i+1] == k {
+			if i < len(set)-1 && set[i+1] == k {
 				continue
 			}
 			if k <= string(searchKey) {
-				result = append(result, k)
-			}
-		}
-		return result
-	}
-
-	if err := quick.CheckEqual(radixAddAndScan, sliceAddSortAndFilter, nil); err != nil {
-		t.Error(err)
-	}
-}
-
-func TestReverseIterator_SeekReverseLowerBoundFuzzFromNonRoot(t *testing.T) {
-	// Some edge cases are only triggered when seeking from a non-root node,
-	// such as when looking for a key that is larger than the values currently
-	// in the tree.
-	//
-	// When starting from the root, the prefix is empty and so it will always
-	// match the subset of the search key of same length (they are both empty).
-	// The search for the lower bound will then return nil (all keys in the
-	// tree are lower than the search key) and the seek process is cut short.
-	//
-	// But when starting from a non-root node, the prefix is not empty and so
-	// it will require a recursive search for the global maximum in the
-	// sub-tree, which is not needed when starting from the root.
-
-	r := New()
-	set := []string{}
-	var n *Node
-
-	radixAddAndScan := func(newKey, searchKey readableString) []string {
-		r, _, _ = r.Insert([]byte(newKey), nil)
-
-		// Start seeking from the first root child or don't seek yet
-		if len(r.Root().edges) == 0 {
-			return []string{}
-		}
-		n = r.Root().edges[0].node
-
-		// Now iterate the tree from searchKey to the beginning
-		it := n.ReverseIterator()
-		result := []string{}
-		it.SeekReverseLowerBound([]byte(searchKey))
-		for {
-			key, _, ok := it.Previous()
-			if !ok {
-				break
-			}
-			result = append(result, string(key))
-		}
-		return result
-	}
-
-	sliceAddSortAndFilter := func(newKey, searchKey readableString) []string {
-		// Return if the tree doesn't have a non-root node present yet
-		if n == nil {
-			return []string{}
-		}
-
-		// Append the key to the set and re-sort
-		set = append(set, string(newKey))
-		sort.Strings(set)
-
-		result := []string{}
-		for i := len(set) - 1; i >= 0; i-- {
-			k := set[i]
-			// Check this is not a duplicate of the previous value we just included.
-			// Note we don't just store the last string to compare because empty
-			// string is a valid value in the set and makes comparing on the first
-			// iteration awkward.
-			if i < len(set)-2 && set[i+1] == k {
-				continue
-			}
-			if k <= string(searchKey) && k[:len(n.prefix)] <= string(n.prefix) {
 				result = append(result, k)
 			}
 		}
@@ -282,12 +209,17 @@ func TestReverseIterator_SeekLowerBound(t *testing.T) {
 			[]string{"bar"},
 		},
 
-		// Found by fuzz test to hit code that wasn't covered by any other example
+		// Found by fuzz test that hit code that wasn't covered by any other example
 		// here.
 		{
 			[]string{"bdgedcdc", "agcbcaba"},
 			"beefdafg",
 			[]string{"bdgedcdc", "agcbcaba"},
+		},
+		{
+			[]string{"", "acc", "accea", "accgbbb", "b", "bdebfc", "bdfdcbb", "becccc", "bgefcfc", "c", "cab", "cbd", "cgeaff", "cggfbcb", "cggge", "dcgbd", "ddd", "decfd", "dgb", "e", "edaffec", "ee", "eedc", "efafdbd", "eg", "egf", "egfcd", "f", "fggfdad", "g", "gageecc", "ggd"},
+			"adgba",
+			[]string{"accgbbb", "accea", "acc", ""},
 		},
 	}
 

